<a href="https://colab.research.google.com/github/GayFaustine/Faustine-TPSDW/blob/main/Tugas2_Data_Preparation_TPSDW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tugas 2: Data Preparation**

In [17]:
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv('/content/data_tugas 2.csv')
print('Shape dataset:', df.shape)
df.head(10)

Shape dataset: (205, 9)


,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
1,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
2,?,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.6,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.6,66.4,54.3
5,?,audi,gas,two,sedan,99.8,177.3,66.3,53.1
6,158,audi,gas,four,sedan,105.8,192.7,71.4,55.7
7,?,audi,gas,four,wagon,105.8,NaN,71.4,55.7
8,158,audi,gas,four,sedan,105.8,192.7,71.4,55.9
9,?,audi,gas,two,hatchback,99.5,178.2,67.9,52.0


# **1. Menampilkan Jumlah Data Hilang**

In [11]:
# Ganti '?' dengan NaN
df.replace('?', np.nan, inplace=True)

# Tampilkan jumlah missing value setiap kolom
print(df.isnull().sum())

# Variabel dengan missing value terbanyak
print(f"\nVariabel dengan missing value terbanyak: '{df.isnull().sum().idxmax()}' ({df.isnull().sum().max()} missing values)")

normalized-losses    41
make                  4
bahan-bakar          13
jumlah-pintu         18
body-style           12
wheel-bases           3
length                8
width                 5
height                4
dtype: int64

Variabel dengan missing value terbanyak: 'normalized-losses' (41 missing values)


Berdasarkan output di atas variabel normalized-losses yang memiliki data hilang paling banyak dengan jumlah 41 missing values dari total 205 baris data

# **2. Mengganti Missing Value pada Variabel Kategori**

In [12]:
# Identifikasi missing value pada variabel kategorik
kategorik = ['make', 'bahan-bakar', 'jumlah-pintu', 'body-style']
print("Missing value sebelum diganti:")
print(df[kategorik].isnull().sum())

# Ganti missing value dengan 'unknown'
df[kategorik] = df[kategorik].fillna('unknown')

# Tabel ringkasan jumlah setiap kategori
for col in kategorik:
    print(f"\n--- {col} ---")
    print(df[col].value_counts())

Missing value sebelum diganti:
make             4
bahan-bakar     13
jumlah-pintu    18
body-style      12
dtype: int64

--- make ---
make
toyota           32
mazda            17
nissan           17
honda            13
mitsubishi       12
volkswagen       12
volvo            11
subaru           11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           4
unknown           4
isuzu             4
alfa-romero       3
chevrolet         3
jaguar            3
renault           2
mercury           1
Name: count, dtype: int64

--- bahan-bakar ---
bahan-bakar
gas        173
diesel      19
unknown     13
Name: count, dtype: int64

--- jumlah-pintu ---
jumlah-pintu
four       101
two         86
unknown     18
Name: count, dtype: int64

--- body-style ---
body-style
sedan          88
hatchback      67
wagon          24
unknown        12
hardtop         8
convertible     6
Name: count, dtype: int64


Berdasarkan hasil identifikasi di atas, ditemukan missing value pada keempat variabel kategorik yaitu jumlah-pintu (18), bahan-bakar (13), body-style (12), dan make (4), di mana jumlah-pintu lah yang memiliki missing value terbanyak yaitu 18.

# **3. Mengganti Missing Value pada Variabel Numerik**

In [13]:
# Variabel numerik bertipe float
numerik = ['length', 'width', 'height']
df[numerik] = df[numerik].astype(float)

# Hitung nilai rata-rata masing-masing variabel
print("Nilai rata-rata (mean) setiap variabel:")
for col in numerik:
    print(f"{col}: {df[col].mean():.4f}")

# Ganti missing value dengan nilai mean
for col in numerik:
    df[col] = df[col].fillna(df[col].mean())

# Cek missing value setelah imputasi
print("\nMissing value setelah imputasi:")
print(df[numerik].isnull().sum())

Nilai rata-rata (mean) setiap variabel:
length: 173.9025
width: 65.9310
height: 53.7373

Missing value setelah imputasi:
length    0
width     0
height    0
dtype: int64


Berdasarkan output di atas sebelum imputasi, terdapat missing value pada variabel length (8), width (5), dan height (4). Semua data yang hilang tersebut kemudian diganti menggunakan nilai rata-rata (mean) masing-masing variabel, yaitu length sebesar 173.9025, width sebesar 65.9310, dan height sebesar 53.7373. Setelah proses imputasi selesai, seluruh missing value berhasil terisi sehingga jumlah data kosong pada ketiga variabel tersebut menjadi 0, yang artinya data sudah bersih

# **4. Min-Max Normalization**

In [14]:
# Wheel-bases bertipe float
df['wheel-bases'] = df['wheel-bases'].astype(float)

# Min-Max Normalization: (x - min) / (max - min)
df['wheel-bases-normalized'] = (df['wheel-bases'] - df['wheel-bases'].min()) / \
                                (df['wheel-bases'].max() - df['wheel-bases'].min())

# Tampilkan hasil
print(df[['wheel-bases', 'wheel-bases-normalized']].head(10))

   wheel-bases  wheel-bases-normalized
0         88.6                0.058309
1         88.6                0.058309
2         94.5                0.230321
3         99.8                0.384840
4         99.4                0.373178
5         99.8                0.384840
6        105.8                0.559767
7        105.8                0.559767
8        105.8                0.559767
9         99.5                0.376093


Kolom wheel-bases-normalized merupakan hasil Min-Max Normalization dari kolom wheel-bases, di mana setiap nilai diubah ke dalam skala 0 sampai 1. Semakin kecil nilai wheel-bases maka semakin kecil nilai normalized-nya, begitu pula sebaliknya. Contohnya, nilai wheel-bases 88.6 menghasilkan nilai normalized 0.058 (mendekati 0 karena termasuk nilai terkecil), sedangkan nilai 105.8 menghasilkan 0.559 yang berarti berada di tengah rentang data.

# **5. Z-Score Normalization**

In [15]:
# Z-Score Normalization: (x - mean) / std
for col in ['length', 'width', 'height']:
    df[f'{col}-zscore'] = (df[col] - df[col].mean()) / df[col].std()

# Tampilkan hasil
print(df[['length', 'length-zscore', 'width', 'width-zscore', 'height', 'height-zscore']].head(10))

       length  length-zscore  width  width-zscore  height  height-zscore
0  168.800000      -0.428526   64.1     -0.867630    48.8      -2.038094
1  168.800000      -0.428526   64.1     -0.867630    48.8      -2.038094
2  171.200000      -0.226967   65.5     -0.204232    52.4      -0.552035
3  176.600000       0.226541   66.2      0.127467    54.3       0.232274
4  176.600000       0.226541   66.4      0.222238    54.3       0.232274
5  177.300000       0.285329   66.3      0.174853    53.1      -0.263079
6  192.700000       1.578666   71.4      2.591516    55.7       0.810186
7  173.902538       0.000000   71.4      2.591516    55.7       0.810186
8  192.700000       1.578666   71.4      2.591516    55.9       0.892744
9  178.200000       0.360913   67.9      0.933022    52.0      -0.717153


Kolom length-zscore, width-zscore, dan height-zscore merupakan hasil Z-Score Normalization dari masing-masing variabel. Nilai zscore menunjukkan seberapa jauh data tersebut dari nilai rata-ratanya. Nilai positif berarti data berada di atas rata-rata, nilai negatif berarti data berada di bawah rata-rata, dan nilai 0 berarti data tepat sama dengan rata-rata. Contohnya pada baris ke-7, nilai length sebesar 173.90 menghasilkan zscore 0.000 karena nilainya sama persis dengan mean, sedangkan nilai height 48.8 menghasilkan zscore -2.038 yang berarti nilainya jauh di bawah rata-rata.

In [16]:
print('Shape akhir:', df.shape)
df.head(10)

Shape akhir: (205, 13)


,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height,wheel-bases-normalized,length-zscore,width-zscore,height-zscore
0,NaN,alfa-romero,gas,two,convertible,88.6,168.800000,64.1,48.8,0.058309,-0.428526,-0.867630,-2.038094
1,NaN,alfa-romero,gas,two,convertible,88.6,168.800000,64.1,48.8,0.058309,-0.428526,-0.867630,-2.038094
2,NaN,alfa-romero,gas,two,hatchback,94.5,171.200000,65.5,52.4,0.230321,-0.226967,-0.204232,-0.552035
3,164,audi,gas,four,sedan,99.8,176.600000,66.2,54.3,0.384840,0.226541,0.127467,0.232274
4,164,audi,gas,four,sedan,99.4,176.600000,66.4,54.3,0.373178,0.226541,0.222238,0.232274
5,NaN,audi,gas,two,sedan,99.8,177.300000,66.3,53.1,0.384840,0.285329,0.174853,-0.263079
6,158,audi,gas,four,sedan,105.8,192.700000,71.4,55.7,0.559767,1.578666,2.591516,0.810186
7,NaN,audi,gas,four,wagon,105.8,173.902538,71.4,55.7,0.559767,0.000000,2.591516,0.810186
8,158,audi,gas,four,sedan,105.8,192.700000,71.4,55.9,0.559767,1.578666,2.591516,0.892744
9,NaN,audi,gas,two,hatchback,99.5,178.200000,67.9,52.0,0.376093,0.360913,0.933022,-0.717153


Dataset akhir memiliki shape (205 baris, 13 kolom), bertambah dari 9 kolom awal menjadi 13 kolom karena terdapat 4 kolom baru hasil proses normalisasi yaitu wheel-bases-normalized, length-zscore, width-zscore, dan height-zscore. Seluruh tahapan data preparation telah berhasil dilakukan, mulai dari penggantian missing value pada variabel kategorik dengan "unknown", imputasi variabel numerik dengan nilai mean, hingga normalisasi data menggunakan Min-Max dan Z-Score, sehingga dataset sudah bersih dan siap digunakan untuk analisis atau pemodelan machine learning.